src/quemb/molbe/eri_sparse_DF.py

```diff
@@ -981,3 +1129,203 @@ def find_screening_radius(
+    AO_MO_pair_with_offset: list[tuple[int, AOIdx, MOIdx]],
+    AO_reachable_by_MO_with_offset: list[list[tuple[int, AOIdx]]],
+    AO_reachable_by_MO: list[Vector[AOIdx]],
+) -> SemiSparse3DTensor:
+    assert TA.shape[0] == int_mu_nu_P.nao and TA.shape[1] == len(AO_reachable_by_MO)
+    n_unique = len(AO_MO_pair_with_offset)
+    g_unique = np.zeros((n_unique, int_mu_nu_P.naux), dtype=np.float64)
+    keys = np.empty(n_unique, dtype=np.int64)
+
+    # We cannot directly loop as in
+    # ``for offset, mu, i in AO_MO_pair_with_offset```
+    # if we want to parallelise,
+    # hence we need the explicit counter variables.
+    for outer_counter in prange(len(AO_MO_pair_with_offset)):  # type: ignore[attr-defined]
+        offset, mu, i = AO_MO_pair_with_offset[outer_counter]
+        keys[offset] = ravel_Fortran(mu, i, TA.shape[0])
+        for nu_counter in prange(len(int_mu_nu_P.exch_reachable[mu])):  # type: ignore[attr-defined]
+            inner_offset, nu = int_mu_nu_P.exch_reachable_with_offsets[mu][nu_counter]
+            # In an un-optimized, but readable way it would be written as:
+            # g_unique[offset] += TA[nu, i] * int_mu_nu_P[mu, nu]
+            # but we know ahead of time the offset in the `unique_dense_data`
+            # and don't want to compute the lookup.
+            g_unique[offset] += TA[nu, i] * int_mu_nu_P.unique_dense_data[inner_offset]  # type: ignore[index]
+
+    return SemiSparse3DTensor(
+        g_unique,
+        keys,
+        (TA.shape[0], TA.shape[1], int_mu_nu_P.naux),
+        AO_reachable_by_MO_with_offset,
+        AO_reachable_by_MO,
+    )
+
+
+@njit(parallel=True)
+def _second_contract_with_TA(
+    TA: Matrix[np.float64], int_mu_i_P: SemiSparse3DTensor
+) -> Tensor3D[np.float64]:
+    assert TA.shape[0] == int_mu_i_P.shape[0]
+    assert TA.shape[1] == int_mu_i_P.shape[1]
+
+    g = np.zeros((TA.shape[1], TA.shape[1], int_mu_i_P.naux), dtype=np.float64)
+
+    for i in prange(g.shape[0]):  # type: ignore[attr-defined]
+        for j in prange(min(g.shape[1], i + 1)):  # type: ignore[attr-defined]
+            for offset, nu in int_mu_i_P.AO_reachable_by_MO_with_offsets[i]:
+                g[i, j, :] += TA[nu, j] * int_mu_i_P.dense_data[offset]
+
+            g[j, i, :] = g[i, j, :]
+    return g
+
+
+def transform_sparse_DF_integral(
+    mf: scf.hf.SCF,
+    Fobjs: Sequence[Frags],
+    file_eri_handler: h5py.File,
+    auxbasis: str | None = None,
+    screen_radius: Mapping[str, float] | None = None,
+) -> None:
+    eris = _transform_sparse_DF_integral(
+        mf,
+        Fobjs,
+        auxbasis,
+        screen_radius,
+    )
+    write_eris(Fobjs, eris, file_eri_handler)
+
+
+def _slow_transform_sparse_DF_integral(
+    mf: scf.hf.SCF,
+    Fobjs: Sequence[Frags],
+    auxbasis: str | None = None,
+    screen_radius: Mapping[str, float] | None = None,
+) -> list[Matrix[np.float64]]:
+    """Only exist for reference. Can be deleted soon."""
```

Comment:
I think it is useful to keep this :)

Reply:
in this case I perhaps will even expose it to the outside as `int_transfrom="slow-debug-sparse-DF"` and include it in one of the tests.